File: golang/range_sum_query_2d-immutable.go
package main

// Space : O(row*col)
// Time  : O(row*col)

type NumMatrix struct {
	matrix [][]int
}

func NumMatrixConstructor(matrix [][]int) NumMatrix {
	row, col := len(matrix), len(matrix[0])

	for y := 0; y < row; y++ {
		for x := 1; x < col; x++ {
			matrix[y][x] += matrix[y][x-1]
		}
	}

	return NumMatrix{matrix: matrix}
}

func (obj *NumMatrix) SumRegion(row1 int, col1 int, row2 int, col2 int) int {
	ans := 0
	for y := row1; y <= row2; y++ {
		if col1-1 >= 0 {
			ans += (obj.matrix[y][col2] - obj.matrix[y][col1-1])
		} else {
			ans += obj.matrix[y][col2]
		}

	}

	return ans
}

File: python3/range_sum_query_2d_immutable.py
"""
Space   : O(mn)
Time    : O(mn)
"""


class NumMatrix:

    def __init__(self, matrix: List[List[int]]):
        self.matrix = matrix
        row, col = len(matrix), len(matrix[0])
        for y in range(row):
            for x in range(1, col):
                self.matrix[y][x] += self.matrix[y][x-1]

    def sumRegion(self, row1: int, col1: int, row2: int, col2: int) -> int:
        ans = 0
        for y in range(row1, row2+1):
            if col1 == 0:
                ans += self.matrix[y][col2]
            else:
                ans += self.matrix[y][col2] - self.matrix[y][col1-1]

        return ans


# Your NumMatrix object will be instantiated and called as such:
# obj = NumMatrix(matrix)
# param_1 = obj.sumRegion(row1,col1,row2,col2)
